mjpc/trajectory.cc

```diff
@@ -204,6 +200,119 @@ void Trajectory::Rollout(
   UpdateReturn(task);
 }
 
+// simulate model forward in time with discrete-time indexed policy
+void Trajectory::RolloutDiscrete(
+    std::function<void(double* action, const double* state, int index)>
+        policy,
+    const Task* task, const mjModel* model, mjData* data, const double* state,
+    double time, const double* mocap, int steps) {
+  // horizon
+  horizon = steps;
+
+  // set mocap
+  for (int i = 0; i < model->nmocap; i++) {
+    mju_copy(data->mocap_pos + 3 * i, mocap + 7 * i, 3);
+    mju_copy(data->mocap_quat + 4 * i, mocap + 7 * i + 3, 4);
+  }
+
+  // set initial state
+  mju_copy(states.data(), state, dim_state);
+  mju_copy(data->qpos, state, model->nq);
+  mju_copy(data->qvel, state + model->nq, model->nv);
+  mju_copy(data->act, state + model->nq + model->nv, model->na);
+
+  // set initial time
+  times[0] = time;
+  data->time = time;
+
+  // step1
+  mj_step1(model, data);
+
+  // action from policy
+  policy(actions.data(), states.data(), 0);
+  mju_copy(data->ctrl, actions.data(), model->nu);
+
+  // initial residual
+  task->Residuals(model, data, residual.data());
+
+  // initial trace
+  double* tr = SensorByName(model, data, "trace");
+  if (tr) mju_copy(trace.data(), tr, 3);
+
+  for (int t = 1; t < horizon - 1; t++) {
+    // step2
+    mj_step2(model, data);
+
+    // check for step warnings
+    if (CheckWarnings(data)) {
```

Comment:
Can you make sure that the logic to update `failure` is as in the other calls to CheckWarnings, above?

Also reset `failure` to `false` at the start of the rollout.